test/integration/filters/test_socket_interface.cc

```diff
@@ -0,0 +1,79 @@
+#include "test/integration/filters/test_socket_interface.h"
+
+#include <algorithm>
+
+#include "envoy/common/exception.h"
+#include "envoy/extensions/network/socket_interface/v3/default_socket_interface.pb.h"
+#include "envoy/network/socket.h"
+
+#include "common/api/os_sys_calls_impl.h"
+#include "common/common/utility.h"
+#include "common/network/address_impl.h"
+
+namespace Envoy {
+namespace Network {
+
+Api::IoCallUint64Result TestIoSocketHandle::writev(const Buffer::RawSlice* slices,
+                                                   uint64_t num_slice) {
+  if (writev_override_) {
+    auto result = writev_override_(this, slices, num_slice);
+    if (result.has_value()) {
+      return std::move(result).value();
+    }
+  }
+  return IoSocketHandleImpl::writev(slices, num_slice);
+}
+
+IoHandlePtr TestIoSocketHandle::accept(struct sockaddr* addr, socklen_t* addrlen) {
+  auto result = Api::OsSysCallsSingleton::get().accept(fd_, addr, addrlen);
+  if (SOCKET_INVALID(result.rc_)) {
+    return nullptr;
+  }
+
+  return std::make_unique<TestIoSocketHandle>(writev_override_, result.rc_, socket_v6only_);
+}
+
+IoHandlePtr TestSocketInterface::socket(Socket::Type socket_type, Address::Type addr_type,
```

Comment:
I guess this is the result of me making all the function calls const. Are you okay with the approach or should we consider doing things differently?

Reply:
This seems like a ton of copying also that is needed to override. Can we potentially copy less here somehow?

Reply:
Apart from removing the const all over the place, probably the other option would be to keep an internal `SocketInterfaceImpl` instance for `socket()` implementations and only derive `SocketInterface`. Just a thought.

Reply:
I've reworked it to minimize code duplication.